### PR TITLE
EndpointSlice and Endpoints behavior for terminating pods should be the same

### PR DESCRIPTION
**Pull request by andrewsykim**
Signed-off-by: Andrew Sy Kim <kim.andrewsy@gmail.com>

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Updates endpointslice controller to treat terminating pods for EndpointSlice the same as Endpoints. The expected behavior for terminating pods based on Endpoints is that a terminating Pod is only included if `service.Spec.PublishNotReadyAddresses` is true  

**Which issue(s) this PR fixes**:
Based on a discussion with @robscott @lbernail  and @thockin https://github.com/kubernetes/kubernetes/pull/88590#discussion_r390379302 

**Special notes for your reviewer**:
There's some ongoing discussion on how we should treat terminating pods in Endpoints and EndpointSlice in https://github.com/kubernetes/enhancements/pull/1607. Until we also change the behavior in Endpoints, we should keep the two consistent. 

**Does this PR introduce a user-facing change?**:

```release-note
EndpointSlice should not contain endpoints for terminating pods
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/controller/util/endpoint/controller_utils.go [hunk 1]
```diff
@@ -126,18 +126,14 @@ func DeepHashObjectToString(objectToWrite interface{}) string {
...
 // ShouldPodBeInEndpoints returns true if a specified pod should be in an
-// endpoints object.
-func ShouldPodBeInEndpoints(pod *v1.Pod) bool {
+// endpoints object. Terminating pods are only included if publishNotReady is true.
+func ShouldPodBeInEndpoints(pod *v1.Pod, publishNotReady bool) bool {
 	if len(pod.Status.PodIP) == 0 && len(pod.Status.PodIPs) == 0 {
 		return false
 	}
 
-	if pod.Spec.RestartPolicy == v1.RestartPolicyNever {
```
Comment (andrewsykim) [inline pkg/controller/util/endpoint/controller_utils.go]: oops my bad, we need to keep these actually

**Comment by robscott** [comment]
Hey @andrewsykim, thanks for this PR! This is a great bug fix that should really be cherry picked back into 1.17 and 1.18.

It took me awhile to remember the background here. Initially EndpointSlices intentionally didn't support `publishNotReadyAddresses` since the godoc for that field only talks about DNS implementations. When we decided to add it for the sake of compatibility with Endpoints, my PR (https://github.com/kubernetes/kubernetes/pull/84573) missed replicating a [key if statement](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpoint/endpoints_controller.go#L420-L423) from the Endpoints controller. 

Thanks for catching this!

/lgtm
/assign @freehan

**Review by freehan — COMMENTED** [review]
Good catch! Thank you!

/lgtm
/approve

**Comment by andrewsykim** [comment]
/milestone v1.18

Putting this in the v1.18 milestone based on @robscott's comment here https://github.com/kubernetes/kubernetes/pull/89056#issuecomment-597790562

**Pull request merged**